http.go

```diff
@@ -106,6 +116,45 @@ func InstrumentHandlerFunc(label string, f func(http.ResponseWriter, *http.Reque
 	return InstrumentHandler(label, http.HandlerFunc(f))
 }
 
+// traceSampler is a time-based request sampler for tracing.
+//
+// It allows at most one request to be traced during each time interval.
+type traceSampler struct {
+	interval time.Duration
+	rng      *rand.Rand
+
+	mu             sync.Mutex
+	nextSampleTime time.Time
+}
+
+// newTraceSampler returns a new traceSampler that allows at most one request
+// to be traced during each time interval.
+func newTraceSampler(interval time.Duration) *traceSampler {
+	now := time.Now()
+	return &traceSampler{
+		interval:       interval,
+		rng:            rand.New(rand.NewSource(now.UnixNano())),
+		nextSampleTime: now,
+	}
+}
+
+// shouldTrace returns true iff a request should be traced.
+func (s *traceSampler) shouldTrace() bool {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	now := time.Now()
```

Comment:
Performance tweak -- may not matter:

We are potentially causing cache line contention here by locking the mutex on every request, even non-traced ones. An alternative would be to store the nextSampleTime in an atomic int and then change the initial check to not be with a lock:

```
   now := time.Now().UnixNano()
   if now < s.nextSampleTime.Load() {
      return
   }
```

Should check again after acquiring the mutex in case another goroutine has changed nextSampleTime. Not sure if this is worthwhile.

But at the very least, hoist the time.Now() call so it happens before the mutex is acquired.

Reply:
Nice, done!